### ISSUE TITLE
[RFC] Use the Catalog to source Kubernetes clusters instead of `clusterLocator`

### ISSUE THREAD

**Issue by mclarke47** [backstage/backstage#13727]
**Status:** Open for comments

@backjo recently added the excellent addition to the kubernetes plugin where you could source EKS clusters through the Backstage catalog itself.

This would include moving the logic from the GKE, config, localKubeProxy cluster locators.

This way we could use the features that are built into the catalog, create relationships between components/clusters and possibly create pages for the clusters themselves

**Comment by jpeach** [comment]
I would use this if it was available. Having the clusters in the catalog as resources is a good opportunity to link to the owning team, as well as to link dashboard, pagerduty and so-on.

**Comment by jamieklassen** [comment]
@jpeach when you say "if it was available", what are you imagining? Having the catalog plugin include some custom pages for Resources of type `kubernetes-cluster`? Because the [`catalog` clusterLocator](https://backstage.io/docs/features/kubernetes/configuration#catalog) already exists:

https://github.com/backstage/backstage/blob/4e1a0ad45d031bb92e2bab3732e11eba3a4e8a1d/plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts#L36-L64

If I've understood @mclarke47's intention correctly, this discussion is about making a breaking change where effectively `catalog` becomes the only clusterLocator and the other methods are deprecated and removed?

**Comment by jpeach** [comment]
Oh I skimmed over that line in the docs :)

What is ANNOTATION_KUBERNETES_AUTH_PROVIDER for?

**Comment by jamieklassen** [comment]
@jpeach when you add a cluster to your catalog, the value of the `kubernetes.io/auth-provider` annotation,

https://github.com/backstage/backstage/blob/51af8361dea2b8a08170ccafa51b7c7396ce9fdf/packages/catalog-model/src/entity/constants.ts#L53-L59

determines the [`authProvider` value](https://backstage.io/docs/features/kubernetes/configuration#clustersauthprovider) (e.g.  `serviceAccount`, `google`, `aws`, `azure`, etc) that will be used by Backstage when determining how to decorate requests to that cluster with auth metadata.

**Comment by namco1992** [comment]
Hi @mclarke47 @jamieklassen,

Just to chime in, I noticed that when using the catalog locator, the service account token is not set in the `clusterDetails` if we set the `authProvider` to `serviceAccount`. For other auth providers, there are the corresponding `KubernetesAuthTranslator` to [decorate the `clusterDetails`](https://github.com/backstage/backstage/blob/d1c04da5d0/plugins/kubernetes-backend/src/service/KubernetesFanOutHandler.ts#L232) with a service account token. However, it's an no-op when `authProvider=serviceAccount` since it expects the token to be set in the config.

Then later in the `KubernetesClientProvider`, it will fallback to the `loadFromDefault` method, which goes a bit out of hand from there. [`loadFromDefault`](https://github.com/kubernetes-client/javascript/blob/4ea2b0f60ba724eee1f2ab1485e39e9ff1390507/src/config.ts#L298) tries to find your local kubectl config or auto mounted service account (`/var/run/secrets/kubernetes.io/serviceaccount`) , which might not be available in production environment and we have no control over there.
https://github.com/backstage/backstage/blob/d1c04da5d04b3fcbd80135034a823feed6af4884/plugins/kubernetes-backend/src/service/KubernetesClientProvider.ts#L47-L57

AFAICT, the catalog locator doesn't really work well when `authProvider=serviceAccount`, at least the control of where to read the token is lacking. It's weird to require the kube config file to be present in production environment, and loading token from auto mounted service account path (`/var/run/secrets/kubernetes.io/serviceaccount`) won't work if you have multiple clusters, since you would need multiple service account tokens.

**Comment by jamieklassen** [comment]
hey @namco1992 that's almost definitely an oversight in the implementation of the `catalog` clusterLocator that probably merits a separate issue -- there could be a separate annotation for storing the serviceaccounttoken on the entity itself, or we could have a massive bikeshedding conversation about storing secrets in the catalog!

**Comment by mclarke47** [comment]
Perhaps some guidance from the core maintainers about any other examples of secrets being stored beside catalog components would be good @backstage/catalog-core

**Comment by freben** [comment]
Oh for sure no secrets _in_ the catalog, that's just opening them up to be seen by anybody anywhere. It'd have to be passed one way or the other as config to the backend plugin instead (maybe a cluster-to-token mapping or whatever makes sense in this context)

**Comment by namco1992** [comment]
Yes I agree with @freben that we should avoid actually storing the secrets in the catalog. One approach could be replacing the current `NoopKubernetesAuthTranslator` for the service account with something else that can map the cluster to the service account token.
https://github.com/backstage/backstage/blob/93fcda09ced3f96b52f462291d8eafd1be72437e/plugins/kubernetes-backend/src/kubernetes-auth-translator/KubernetesAuthTranslatorGenerator.ts#L47-L49

Maybe we can also support reading from `clusters` if the type is `catalog`, and map the token to the cluster by name:
```yaml
kubernetes:
  serviceLocatorMethod:
    type: 'multiTenant'
  clusterLocatorMethods:
    - type: 'catalog'
      clusters:
        - name: minikube
          serviceAccountToken: xxx
```

However, it might be a bit messy to make it compatible with both the `ConfigClusterLocator` and `CatalogClusterLocator`, and I would say the development experience is not optimal since you have to update the Backstage config when you have new Kubernetes cluster resources added.

**Comment by jamieklassen** [comment]
@namco1992 this RFC does involve deprecating the config clusterlocator so you might not have to worry about compatibility.

Storing a big mapping of clusters (in the catalog) to serviceaccounttokens in the app-config does feel pretty unpleasant -- to muddy the discussion further, what if this RFC was also about deprecating the serviceaccounttoken auth provider? If you're going to have clusters in the catalog (potentially managed by automated ingestion workflows), maybe you also need to have a cleaner way of authenticating with them besides a static out-of-band-provisioned token?

**Comment by namco1992** [comment]
> this RFC does involve deprecating the config clusterlocator so you might not have to worry about compatibility.

I see, but I wasn't expecting to eliminate the config cluster locator (or any other locators) in one future version, instead, I would prefer that we mark the current locators as deprecated and slowly fade out, which means that we would support both ways for a while. Similar to what we did in https://github.com/backstage/backstage/issues/10183. We run our on-prem kubernetes clusters, and we use the service account token, I can foresee that it would be disruptive if we just drop it in one future version.

> to muddy the discussion further, what if this RFC was also about deprecating the serviceaccounttoken auth provider?

I think it's a wider topic that involves not only the Kubernetes auth providers, but also the Backstage auth providers, and probably https://github.com/backstage/backstage/pull/7761? TBH I don't know enough to give any advice here, but I feel it might deserve to be a separate RFC.

Besides, I noticed that we support [custom cluster discovery](https://backstage.io/docs/features/kubernetes/installation#custom-cluster-discovery) currently, and we might want to consider the impact on users who adopted this approach since this way will most likely be gone too.

**Comment by backjo** [comment]
I wonder if the easiest solution here is to let clients plug in a custom AuthTranslator - allowing them to implement whatever custom logic they need to resolve the credentials. This may be something as simple as looking up a serviceAccountToken from the backstage config and adding it to the cluster - or a more complex setup such as generating/presenting a short-lived x509 credential. 

For what it's worth, I agree that it does't make sense for most larger organizations to integrate with the kubernetes functionality using the configuration method at all - but there is some goodness here today in allowing developers / organizations to try out Backstage and the Kubernetes functionality quickly.

**Comment by backjo** [comment]
I wonder if the easiest solution here is to let clients plug in a custom AuthTranslator - allowing them to implement whatever custom logic they need to resolve the credentials. This may be something as simple as looking up a serviceAccountToken from the backstage config and adding it to the cluster - or a more complex setup such as generating/presenting a short-lived x509 credential. 

For what it's worth, I agree that it does't make sense for most larger organizations to integrate with the kubernetes functionality using the configuration method at all - but there is some goodness here today in allowing developers / organizations to try out Backstage and the Kubernetes functionality quickly.

**Comment by mclarke47** [comment]
I think in the short term we should move the client side auth cluster locators to the catalog method deprecate the GKE/AWS lookup cluster locators and keep the config cluster locator until we have a good secrets solution. I think there are some great suggestions here but maybe we just need to write out the requirements for the server side auth solution, i.e:

* Secrets are stored in a safe place not accessible by Backstage users
* Secrets are configurable per cluster

etc...

**Comment by mclarke47** [comment]
As a bit of an aside, I'm wondering if eventually we can leverage relationships between entities in the catalog to store information such as which cluster a service runs on, this could then be a `serviceLocator`. This would hugely reduce response times/number of fetches for users with a large number of clusters.
